R/as_model_out_df.R

```diff
@@ -0,0 +1,211 @@
+    model_id_col <- validate_col_input(model_id_col, call = call)
+    tbl <- rename_col(tbl, model_id_col, old_names, call)
+  }
+  if (!is.null(output_type_col)) {
+    output_type_col <- validate_col_input(output_type_col, call = call)
+    tbl <- rename_col(tbl, output_type_col, old_names, call)
+  }
+  if (!is.null(output_type_id_col)) {
+    output_type_id_col <- validate_col_input(output_type_id_col, call = call)
+    tbl <- rename_col(tbl, output_type_id_col, old_names, call)
+  }
+  if (!is.null(value_col)) {
+    value_col <- validate_col_input(value_col, call = call)
+    tbl <- rename_col(tbl, value_col, old_names, call)
+  }
+  return(tbl)
+}
+
+validate_col_input <- function(x, call = rlang::caller_env()) {
+  if (!is.character(x)) {
+    cli::cli_abort("{.arg {rlang::caller_arg(x)}} must be a {.cls character} string
+                       instead of {.cls {typeof(x)}}",
+      call = call
+    )
+  }
+  if (length(x) > 1L) {
+    cli::cli_warn("{.arg {rlang::caller_arg(x)}} must be character vector of
+                       length {.val {1L}} not length {.val {length(x)}}.
+                      Only element 1 ({.val {x[1]}}) used, the rest ignored.",
+      call = call
+    )
+    x <- x[1]
+  }
+  return(x)
+}
+
+rename_col <- function(x, col_name, old_names, call) {
+  arg_name <- rlang::caller_arg(col_name)
+  if (!col_name %in% names(x)) {
+    cli::cli_abort(c(
+      "x" = "{.arg {arg_name}} value {.val {col_name}}
+                       not a valid column name in {.arg x}",
+      "!" = "Must be one of {.val {old_names}}"
+    ),
+    call = call
+    )
+  } else {
+    new_col_name <- gsub("_col", "", arg_name)
+    names(x)[names(x) == col_name] <- new_col_name
+  }
+  return(x)
+}
+
+trim_tbl_to_task_ids <- function(tbl, task_id_cols, hub_con,
+                                 call = rlang::caller_env()) {
+  if (is.null(task_id_cols) && is.null(hub_con)) {
+    cli::cli_abort(
+      c(
+        "x" = "Cannot trim to task IDs when both
+                  {.arg task_id_cols} and {.arg hub_con} are NULL",
+        "!" = "Please supply appropriate input to either
+                {.arg task_id_cols} or {.arg hub_con}."
+      ),
+      call = call
+    )
+  } else {
+    if (!is.null(task_id_cols)) {
+      if (!all(task_id_cols %in% names(tbl))) {
+        cli::cli_alert_warning(
+          "{.arg task_id_cols} value{?s} {.val
+                  {task_id_cols[!task_id_cols %in% names(tbl)]}
+                  } not a valid {.arg tbl} column name. Ignored."
+        )
+      }
+      task_id_cols <- task_id_cols[task_id_cols %in% names(tbl)]
+    } else {
+      if (!inherits(hub_con, "hub_connection")) {
+        cli::cli_abort("{.arg hub_con} must be a valid object of
+                                   class {.cls hub_connection}",
+          call = call
+        )
+      }
+      task_id_cols <- get_task_id_names(attr(hub_con, "config_tasks"))
+    }
+  }
+
+  return(tbl[, c(task_id_cols, std_colnames)])
```

Comment:
Should we already grab things in the standard order defined in `std_col_order_model_out_tbl` here?  would be nice if `std_col_order_model_out_tbl` had less (or no) work to do later.

Reply:
As `trim_to_task_ids` will more likely than not be `FALSE` in most cases, I feel it better to just do the ordering once at the end once all the trimming has been completed (including removing any empty columns).

Reply:
sounds good, especially given discussion below about the low cost of re-ordering columns :)